### PR TITLE
feature(dspy): Add Milvus in Retrieve

### PR DESCRIPTION
**Pull request by jaelgu**
Add MilvusRM to support retrieval for Milvus or Zilliz Cloud:
- milvus: https://github.com/milvus-io/milvus
- Zilliz Cloud: https://zilliz.com

### CODE REVIEW

**Comment by arnavsinghvi11** [comment]
Thanks @jaelgu ! Looks great! Could you also add some documentation mirroring some of the current Retrieval Model Client integrations? (e.g. for [ChromadbRM](https://dspy-docs.vercel.app/api/retrieval_model_clients/ChromadbRM))

Will merge the PR after that!

**Comment by jaelgu** [comment]
> Thanks @jaelgu ! Looks great!
> 
> Looks like couple of the tests are failing. Can you run:
> 
> 1. `ruff check . --fix-only`
> 2. `poetry install --no-interaction`
>    Seeing this error:
> 
> ```
> Installing dependencies from lock file
> Warning: poetry.lock is not consistent with pyproject.toml. You may be getting improper dependencies. Run `poetry lock [--no-update]` to fix it.
> 
> Because dspy depends on pymilvus (^2.3.7) which doesn't match any versions, version solving failed.
> ```
> 
> Could you also add some documentation mirroring some of the current Retrieval Model Client integrations? (e.g. for [ChromadbRM](https://dspy-docs.vercel.app/api/retrieval_model_clients/ChromadbRM))
> 
> Will merge the PR after that!

Hi, thank you for feedback. I've followed instructions to update codes and added docs.

**Comment by arnavsinghvi11** [comment]
Hi @jaelgu , thanks for the additions. 

The PR is failing the poetry.lock test with this error:

```
Run poetry install --no-interaction
Installing dependencies from lock file
Warning: poetry.lock is not consistent with pyproject.toml. You may be getting improper dependencies. Run `poetry lock [--no-update]` to fix it.

Because dspy depends on pymilvus (^2.3.6) which doesn't match any versions, version solving failed.
```

Can you rebase this PR to the main branch, and then run `poetry lock --no-update` and commit the changed `poetry.lock` file? It is key to rebase to the latest version on main since the `poetry.lock` hash and updates need to be maintained. lmk if you have any questions on this.

**Comment by jaelgu** [comment]
> Hi @jaelgu , thanks for the additions.
> 
> The PR is failing the poetry.lock test with this error:
> 
> ```
> Run poetry install --no-interaction
> Installing dependencies from lock file
> Warning: poetry.lock is not consistent with pyproject.toml. You may be getting improper dependencies. Run `poetry lock [--no-update]` to fix it.
> 
> Because dspy depends on pymilvus (^2.3.6) which doesn't match any versions, version solving failed.
> ```
> 
> Can you rebase this PR to the main branch, and then run `poetry lock --no-update` and commit the changed `poetry.lock` file? It is key to rebase to the latest version on main since the `poetry.lock` hash and updates need to be maintained. lmk if you have any questions on this.

Hello, thank you for help! I've updated the `poetry.lock` after running `poetry lock --no-update` at local.

File: poetry.lock [hunk 1]
```diff
@@ -1906,6 +2023,20 @@ qtconsole = ["qtconsole"]
 test = ["pickleshare", "pytest (<7.1)", "pytest-asyncio (<0.22)", "testpath"]
 test-extra = ["curio", "matplotlib (!=3.2.0)", "nbformat", "numpy (>=1.22)", "pandas", "pickleshare", "pytest (<7.1)", "pytest-asyncio (<0.22)", "testpath", "trio"]
 
+[[package]]
+name = "isodate"
```
Comment (arnavsinghvi11) [inline poetry.lock:2027]: same comment as above

File: poetry.lock [hunk 2]
```diff
@@ -3932,6 +4141,32 @@ pyyaml = "*"
 [package.extras]
 extra = ["pygments (>=2.12)"]
 
+[[package]]
+name = "pymilvus"
```
Comment (arnavsinghvi11) [inline poetry.lock:4145]: this one is good to keep :)

**Review by arnavsinghvi11 — CHANGES_REQUESTED** [review]

**Comment by arnavsinghvi11** [comment]
Thanks @jaelgu ! looks good to merge, pending the checks on the extra dependencies added to poetry.lock.

File: poetry.lock [hunk 3]
```diff
@@ -208,6 +208,63 @@ files = [
     {file = "appnope-0.1.4.tar.gz", hash = "sha256:1de3860566df9caf38f01f86f65e0e13e379af54f9e4bee1e66b48f2efffd1ee"},
 ]
 
+[[package]]
+name = "argon2-cffi"
```
[the remark this answers is not in the document]
Comment (jaelgu) [reply poetry.lock:212]: Yes this is required by pymilvus. Pymilvus requires "minio", which requires "argon2-cffi".

File: poetry.lock [hunk 4]
```diff
@@ -302,6 +359,45 @@ docs = ["sphinx-copybutton (>=0.4,<0.5)", "sphinx-rtd-theme (>=1.0,<2.0)", "sphi
 erdantic = ["erdantic (>=0.6,<0.7)"]
 test = ["coverage (>=7,<8)", "pytest (>=7,<8)"]
 
+[[package]]
+name = "azure-core"
```
[the remark this answers is not in the document]
Comment (jaelgu) [reply poetry.lock:363]: pymilvus requires "azure-storage-blob", which requires "azure-core".

File: poetry.lock [hunk 5]
```diff
@@ -1005,6 +1101,27 @@ files = [
     {file = "docutils-0.16.tar.gz", hash = "sha256:c2de3a60e9e7d07be26b7f2b00ca0309c207e06c100f9cc2a94931fc75a478fc"},
 ]
 
+[[package]]
+name = "environs"
```
[the remark this answers is not in the document]
Comment (jaelgu) [reply poetry.lock:1105]: yes, "environs" is required by pymilvus.
image

File: poetry.lock (hunk 1 above)
Comment (jaelgu) [reply poetry.lock:2027]: pymilvus requires "azure-storage-blob", which requires "isodate".

File: poetry.lock [hunk 6]
```diff
@@ -2278,6 +2409,25 @@ pydantic = "*"
 requests = "*"
 urllib3 = "*"
 
+[[package]]
+name = "marshmallow"
```
[the remark this answers is not in the document]
Comment (jaelgu) [reply poetry.lock:2413]: pymilvus requires "environs", which requires "mashmallow".

File: poetry.lock [hunk 7]
```diff
@@ -2357,6 +2507,24 @@ verspec = "*"
 dev = ["coverage", "flake8 (>=3.0)", "flake8-quotes", "shtab"]
 test = ["coverage", "flake8 (>=3.0)", "flake8-quotes", "shtab"]
 
+[[package]]
+name = "minio"
```
[the remark this answers is not in the document]
Comment (jaelgu) [reply poetry.lock:2511]: yes, "minio" is required by pymilvus.

File: poetry.lock [hunk 8]
```diff
@@ -3770,6 +3938,47 @@ files = [
     {file = "pycparser-2.21.tar.gz", hash = "sha256:e644fdec12f7872f86c58ff790da456218b10f863970249516d60a5eaca77206"},
 ]
 
+[[package]]
+name = "pycryptodome"
```
[the remark this answers is not in the document]
Comment (jaelgu) [reply poetry.lock:3942]: pymilvus requires "minio", which requires "pycrytodome".

**Comment by jaelgu** [comment]
> Thanks @jaelgu ! looks good to merge, pending the checks on the extra dependencies added to poetry.lock.

I've replied to the comments and it turns out all extra dependencies are required by pymilvus.

**Comment by arnavsinghvi11** [comment]
Thanks @jaelgu for checking!

**Pull request merged**